### PR TITLE
Update old cabal examples to version 3.0

### PR DESCRIPTION
**Pull request by sekunho**
The examples contain pre-SPDX identifiers which causes `cabal check`
to complain about it being an unknown SPDX license identifier.

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by Mikolaj** [comment]
Whoa, we have 284 occurences of `BSD3` in the code. I guess nothing else than `check` checks that (and it's only a warning?).

**Review by Mikolaj — APPROVED** [review]

**Comment by sekunho** [comment]
Yeah I wasn't sure if I should modify the ones outside those examples, so I just left them alone for now. If someone could help point me to which ones are OK to modify then I'll gladly work on them too.

**Comment by Mikolaj** [comment]
Wow, Cabal itself has this wrong licence. 

> Cabal/Cabal.cabal:license:       BSD3

```
~/r/cabal$ cd Cabal
~/r/cabal/Cabal$ cabal check
No errors or warnings could be found in the package.
```

How come?

**Comment by Mikolaj** [comment]
I'd guess it's safe to modify all *except* those occurence that test that `cabal check` complains as it should and except the bit of code that suggests to replace BSD3 with BSD-3-Clause (`templates/SPDX.LicenseId.template.hs`) and except `Cabal/src/Distribution/PackageDescription/Quirks.hs` that should be just left alone.

**Comment by Mikolaj** [comment]
And `Cabal/src/Distribution/SPDX/LicenseId.hs` and `Cabal/src/Distribution/PackageDescription/Check.hs`.

**Comment by Mikolaj** [comment]
The latter probably explains whey `cabal check` doesn't complain, but I'm curious how you made it complain.

**Comment by Mikolaj** [comment]
Oh, you probably have some ancient `cabal-version` specified in your .cabal file?

**Comment by sekunho** [comment]
I made a tiny reproducible example here: https://github.com/sekunho/cabal-check-fail

```
[sekun@nixos:~/Projects/cabal-check-fail]$ cabal --version
cabal-install version 3.6.2.0
compiled using version 3.6.2.0 of the Cabal library 

[sekun@nixos:~/Projects/cabal-check-fail]$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.10.7
```

`cabal` file

```
cabal-version:      2.4
name:               cabal-check-fail
version:            0.1.0.0
synopsis:

-- A longer description of the package.
-- description:
homepage:

-- A URL where users can report bugs.
-- bug-reports:
license:            BSD3
license-file:       LICENSE
author:             SEKUN
maintainer:         sekun@hey.com

-- A copyright notice.
-- copyright:
-- category:
extra-source-files: CHANGELOG.md

library
    exposed-modules:  MyLib

    -- Modules included in this library but not exported.
    -- other-modules:

    -- LANGUAGE extensions used by modules in this package.
    -- other-extensions:
    build-depends:    base ^>=4.14.3.0
    hs-source-dirs:   src
    default-language: Haskell2010

executable cabal-check-fail
    main-is:          Main.hs

    -- Modules included in this executable, other than Main.
    -- other-modules:

    -- LANGUAGE extensions used by modules in this package.
    -- other-extensions:
    build-depends:
        base ^>=4.14.3.0,
        cabal-check-fail

    hs-source-dirs:   app
    default-language: Haskell2010

test-suite cabal-check-fail-test
    default-language: Haskell2010
    type:             exitcode-stdio-1.0
    hs-source-dirs:   test
    main-is:          MyLibTest.hs
    build-depends:    base ^>=4.14.3.0
```

I chose a version of cabal after SPDX was introduced.

**Comment by sekunho** [comment]
I made a tiny reproducible example here: https://github.com/sekunho/cabal-check-fail

```
[sekun@nixos:~/Projects/cabal-check-fail]$ cabal --version
cabal-install version 3.6.2.0
compiled using version 3.6.2.0 of the Cabal library 

[sekun@nixos:~/Projects/cabal-check-fail]$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.10.7
```

`cabal` file

```
cabal-version:      2.4
name:               cabal-check-fail
version:            0.1.0.0
synopsis:

-- A longer description of the package.
-- description:
homepage:

-- A URL where users can report bugs.
-- bug-reports:
license:            BSD3
license-file:       LICENSE
author:             SEKUN
maintainer:         john@example.com

-- A copyright notice.
-- copyright:
-- category:
extra-source-files: CHANGELOG.md

library
    exposed-modules:  MyLib

    -- Modules included in this library but not exported.
    -- other-modules:

    -- LANGUAGE extensions used by modules in this package.
    -- other-extensions:
    build-depends:    base ^>=4.14.3.0
    hs-source-dirs:   src
    default-language: Haskell2010

executable cabal-check-fail
    main-is:          Main.hs

    -- Modules included in this executable, other than Main.
    -- other-modules:

    -- LANGUAGE extensions used by modules in this package.
    -- other-extensions:
    build-depends:
        base ^>=4.14.3.0,
        cabal-check-fail

    hs-source-dirs:   app
    default-language: Haskell2010

test-suite cabal-check-fail-test
    default-language: Haskell2010
    type:             exitcode-stdio-1.0
    hs-source-dirs:   test
    main-is:          MyLibTest.hs
    build-depends:    base ^>=4.14.3.0
```

I chose a version of cabal after SPDX was introduced.

**Comment by Mikolaj** [comment]
Oh yes, you are right --- it's Cabal.cabal that has an old cabal-version and that's why `cabal check` doesn't complain.

Wow, and when I update the licence in Cabal.cabal, it complains.

So it's dependent on the cabal-version. 

In that case, I guess it's safe to only change those you proposed at the start. BTW, with what cabal-version would they be run, by default? What is the user most likely to use?

**Comment by Kleidukos** [comment]
@Mikolaj I think version 3 is what people are most likely to use nowadays, and it's compatible with SPDX identifiers

**Review by Kleidukos — APPROVED** [review]
While there may be much more work to do in the future, let's not block this PR

**Comment by Mikolaj** [comment]
Let's ride!

**Review by fgaz — CHANGES_REQUESTED** [review]
> While there may be much more work to do in the future, let's not block this PR

I think the examples should be consistent. @sekunho could you update `cabal-version` in those examples to 3.0 and check that there are no additional warnings?

**Comment by Mikolaj** [comment]
@fgaz has a point: some of these cabal-versions are ancient so new licence formats might not work with them. I do hope other things in these sample configs work fine with new cabal-versions. If not, they would be a bad fit for the manual. so I hope somebody would have spotted that already.

**Comment by ptkato** [comment]
Aye, at first I haven't realised, but for any `.cabal` file with `cabal-version < 2.2`, the examples should remain in their old state, either that or bump the `cabal-version` to some version `>= 2.2`.

**Comment by sekunho** [comment]
> In that case, I guess it's safe to only change those you proposed at the start. BTW, with what cabal-version would they be run with, by default? What is the user most likely to use?

`cabal init --interactive` has `2.4` as the default

```
Please choose version of the Cabal specification to use:
   1) 1.10   (legacy)
   2) 2.0    (+ support for Backpack, internal sub-libs, '^>=' operator)
   3) 2.2    (+ support for 'common', 'elif', redundant commas, SPDX)
 * 4) 2.4    (+ support for '**' globbing)
   5) 3.0    (+ set notation for ==, common stanzas in ifs, more redundant commas, better pkgconfig-depends)
Your choice? [default: 2.4    (+ support for '**' globbing)] 
```

Although I'm not sure what `stack` generates as the default. Should this be the new one in the examples, or do I stick with `2.2`? 

> Sorry for blocking this, but I think the examples should not be left in an inconsistent state. @sekunho could you update cabal-version in those examples to 3.0 (or [any that uses SPDX identifiers](https://cabal.readthedocs.io/en/3.6/file-format-changelog.html)) and check that there are no additional warnings?

Agh I hadn't realized that; thanks for pointing that out. Let me work on it.

**Comment by ptkato** [comment]
> `cabal init --interactive` has `2.4` as the default

By the way, in latest master the default for `cabal init` is `3.0`.

https://github.com/haskell/cabal/blob/55ad63273ce4ed3962dd59d0eca80510dc3ddc87/cabal-install/src/Distribution/Client/Init/Defaults.hs#L70

**Comment by sekunho** [comment]
Alright `3.0` it is. Thanks @ptkato!

**Review by fgaz — CHANGES_REQUESTED** [review]
As i wrote in the last review, please check that the examples do not give more errors/warnings than they did before, using `cabal check`. There currently are parse errors in the `cabal-version` fields.

**Comment by fgaz** [comment]
Hmm, there is still a new warning about missing `default-language` fields... I'm not sure if adding it is worth it or if it's just noise

**Comment by jneira** [comment]
@Mergifyio rebase master

**Comment by jneira** [comment]
@Mergifyio rebase master

**Comment by jneira** [comment]
Removing the merge me label until @fgaz rereview last changes in [d5d381c](https://github.com/haskell/cabal/pull/7942/commits/d5d381cabc178211160288af5bcb0eea8d919ecd)

**Comment by sekunho** [comment]
Ok done. I also added the missing `default-language` field for old examples that gave that warning. There are other warnings about missing fields of descriptions, maintainers, and categories but this was there since before, and I don't know if there are any benefits to adding those in each example.

I summarized the changes in the commit message. Let me know if there's anything else to change/add/remove.

**Comment by Mikolaj** [comment]
Great. So that's ready for @fgaz's review.

**Review by fgaz — APPROVED** [review]
Thanks for all the changes, looks good now!

> There are other warnings about missing fields of descriptions, maintainers, and categories but those were there since before, and I don't know if there are any benefits to adding those in each example.

I agree it's fine. As you wrote, they were already there, and it's only metadata

**Pull request merged**
